src/ui/VirtualFileRenderer/VirtualFileRenderer.tsx

```diff
@@ -0,0 +1,323 @@
+import { useWindowVirtualizer } from '@tanstack/react-virtual'
+// eslint-disable-next-line no-restricted-imports
+import type { Dictionary } from 'lodash'
+import Highlight, { defaultProps } from 'prism-react-renderer'
+import { useEffect, useLayoutEffect, useRef } from 'react'
+import { useHistory, useLocation } from 'react-router-dom'
+
+import { requestAnimationTimeout } from 'shared/utils/animationFrameUtils'
+import { cn } from 'shared/utils/cn'
+import { prismLanguageMapper } from 'shared/utils/prismLanguageMapper'
+
+import { ColorBar } from './ColorBar'
+
+// copied from prism-react-renderer since they don't export it
+type Token = {
+  types: string[]
+  content: string
+  empty?: boolean
+}
+
+interface CodeBodyProps {
+  tokens: Token[][]
+  getLineProps: Highlight['getLineProps']
+  getTokenProps: Highlight['getTokenProps']
+  coverage: Dictionary<'H' | 'M' | 'P'>
+  codeDisplayOverlayRef: React.RefObject<HTMLDivElement>
+}
+
+const CodeBody = ({
+  tokens,
+  getLineProps,
+  getTokenProps,
+  coverage,
+  codeDisplayOverlayRef,
+}: CodeBodyProps) => {
+  const history = useHistory()
+  const location = useLocation()
+
+  const initialRender = useRef(true)
+  const wrapperRef = useRef<HTMLDivElement>(null)
+
+  // virtualize code lines
+  const virtualizer = useWindowVirtualizer({
+    count: tokens.length,
+    estimateSize: () => 18,
+    overscan: 75,
+    scrollMargin: codeDisplayOverlayRef.current?.offsetTop ?? 0,
+  })
+
+  const div = codeDisplayOverlayRef.current
+  if (div) {
+    // set the parent div height to the total size of the virtualizer
+    div.style.height = `${virtualizer.getTotalSize()}px`
+    div.style.position = 'relative'
+  }
+
+  // TODO - finalize this logic
+  // scroll to line on the initial render
+  useEffect(() => {
```

Comment:
Can this be a useLayoutEffect?

Reply:
It can be, however ... it doesn't resolve the issue of the scroll to function not being called before rendering, which I haven't found any "solution" provided via docs or issues in their repo.

Reply:
I think the solution here is to do something like

if loading
  render <LoadingPlaceholder/>
else
  render <VirtualizedList/>
  
That way you should be able to entirely avoid raf and just call it inside a useLayoutEffect as the first render will already contain all the data